### PR TITLE
Add allow server listing client option

### PR DESCRIPTION
**Pull request by NoahvdAa**
Adds the `ALLOWS_SERVER_LISTINGS` client option from #7268. Also addresses [MM's comment](https://github.com/PaperMC/Paper/pull/7268#discussion_r777680170) by adding a constructor that takes a map as an argument. I'm not entirely sure how I feel about this map approach, but it's definitely better than adding a new constructor every time.

### CODE REVIEW

**Review by electronicboy — APPROVED** [review]

File: patches/api/0194-Add-Player-Client-Options-API.patch [hunk 1]
```diff
@@ -127,6 +130,19 @@ index 0000000000000000000000000000000000000000..f7f171c4ee0b8339b2f8fbe82442d65f
 +        this.chatColors = chatColors;
 +        this.skinparts = skinParts;
 +        this.mainHand = mainHand;
++        this.allowsServerListings = false;
++    }
++
++    public PlayerClientOptionsChangeEvent(@NotNull Player player, @NotNull Map<ClientOption<?>, ?> options) {
```
Comment (Machine-Maker) [inline patches/api/0194-Add-Player-Client-Options-API.patch:139]: Add an ApiStatus.Internal annotation to the ctor. no plugin should really need to construct this event I think.

File: patches/api/0194-Add-Player-Client-Options-API.patch [hunk 2]
```diff
@@ -127,6 +130,19 @@ index 0000000000000000000000000000000000000000..f7f171c4ee0b8339b2f8fbe82442d65f
...
++        super(player);
++
++        this.locale = (String) options.get(ClientOption.LOCALE);
++        this.viewDistance = (int) options.get(ClientOption.VIEW_DISTANCE);
++        this.chatVisibility = (ChatVisibility) options.get(ClientOption.CHAT_VISIBILITY);
++        this.chatColors = (boolean) options.get(ClientOption.CHAT_COLORS_ENABLED);
++        this.skinparts = (SkinParts) options.get(ClientOption.SKIN_PARTS);
++        this.mainHand = (MainHand) options.get(ClientOption.MAIN_HAND);
++        this.allowsServerListings = (boolean) options.get(ClientOption.ALLOW_SERVER_LISTINGS);
```
Comment (Machine-Maker) [inline patches/api/0194-Add-Player-Client-Options-API.patch:148]: wrap these in requireNonNulls. yeah, the map isn't great. but its better than ending up with 40 constructors in 2 years I think

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Review by Machine-Maker — APPROVED** [review]

**Comment by NoahvdAa** [comment]
Rebased for 1.19, label can be removed!

**Comment by Machine-Maker** [comment]
is the `textFilteringEnabled` field in the ServerboundClientInformationPacket not used by the client yet? or can that be added here too?. I see it checking a flag on the userApiService on the client, so maybe that's something that can be set on the account?

**Comment by NoahvdAa** [comment]
Could've sworn that was part of the original PR, it has now been added!

**Comment by NoahvdAa** [comment]
Rebased.

**Review by kennytv — APPROVED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Comment by Owen1212055** [comment]
~~Out of the scope of this PR, but PlayerClientOptionsChangeEvent could use similar love.~~ I am blind.

**Review by Owen1212055 — APPROVED** [review]

**Pull request merged**